src/auth/Crypto.cc

```diff
@@ -35,30 +36,58 @@
 #include "common/debug.h"
 #include <errno.h>
 
-int get_random_bytes(char *buf, int len)
+// use a mersenne_twister_engine for random numbers. this engine is faster than
+// std::random_device, and generates higher quality randomness than the other
+// alternatives
+using base_random_engine = std::mt19937;
+
+// wrap with an independent_bits_engine to allow get_random_bytes() to sample
+// a single byte at a time. uniform_int_distribution can't be used for this,
+// because it is undefined for integer types smaller than short
+using streaming_random_engine = std::independent_bits_engine<
+    base_random_engine, 8, unsigned char>;
+
+// engine factory function
+static streaming_random_engine make_seeded_random_engine()
 {
-  int fd = TEMP_FAILURE_RETRY(::open("/dev/urandom", O_RDONLY));
-  if (fd < 0)
-    return -errno;
-  int ret = safe_read_exact(fd, buf, len);
-  VOID_TEMP_FAILURE_RETRY(::close(fd));
-  return ret;
+  // mt19937 has more internal state than can be seeded with a single integer.
+  // use a random_device to produce a seed sequence of 8
+  std::random_device r;
+  std::seed_seq seed{r(), r(), r(), r(), r(), r(), r(), r()};
+  streaming_random_engine engine(seed);
+  return engine;
 }
 
-static int get_random_bytes(int len, bufferlist& bl)
+// return a reference to a singleton engine that's seeded on first use
+static streaming_random_engine& get_seeded_random_engine()
 {
-  char buf[len];
-  get_random_bytes(buf, len);
-  bl.append(buf, len);
-  return 0;
+  static streaming_random_engine engine = make_seeded_random_engine();
+  return engine;
+}
+
+static std::mutex random_mutex;
+
+void get_random_bytes(char *buf, int len)
```

Comment:
I wonder if it would be worthwhile to have a version that just returns a std:array of random bytes when we know the number of bytes needed at compile time.

(I'm tempted to import span from the GSL sometime since it supports using compile time information when we know what it is.)

Reply:
i think it would be cool to have an interface based on iterators so you could use it with any container or raw pointer. for now though, i'd rather limit the changes to call sites

Reply:
+1